include/cadmium/celldevs/cell/cell.hpp

```diff
@@ -0,0 +1,224 @@
+                state.neighbors_vicinity[entry.first] = entry.second;
+                state.neighbors_state[entry.first] = S();
+            }
+            buffer = delayer_factory<T, S>::create_delayer(delayer_id, std::forward<Args>(args)...);
+            buffer->add_to_buffer(initial_state, T());  // At t = 0, every cell communicates its state to its neighbors
+        }
+
+        /************** USER-DEFINED METHODS **************/
+        /// @return cell's next state.
+        virtual S local_computation() const { return state.current_state; }
+        /// @return delay to be applied before communicating to neighbors a new state.
+        virtual T output_delay(S const &cell_state) const { return std::numeric_limits<T>::infinity(); }
+
+        /****************** DEVS METHODS ******************/
+        /// @brief the internal transition function clears output delayer buffer and updates clock and next time advance.
+        void internal_transition() {
+            buffer->pop_buffer();
+            clock += time_advance();
+            next_internal = buffer->next_timeout() - clock;
+        }
+
+        /**
+         * External transition function.
+         * @brief It updates clock and next internal event. Then, it refreshes neighbors' state and computes next cell state.
+         *        if the new cell state is different to the current state, it adds the new state to the output delayer buffer.
+         * @param e elapsed time from the last event.
+         * @param mbs message bag containing new neighbors' state messages.
+         */
+        void external_transition(T e, typename cadmium::make_message_bags<input_ports>::type mbs) {
+            // Update clock and next internal event
+            clock += e;
+            next_internal -= e;
+            // Refresh the neighbors' current state
+            std::vector<cell_state_message<C, S>> bagPortIn = cadmium::get_messages<typename cell_ports_def<C, S>::cell_in>(mbs);
+            for (cell_state_message<C, S> msg: bagPortIn) {
+                auto it = state.neighbors_state.find(msg.cell_id);
+                if (it != state.neighbors_state.end()) {
+                    state.neighbors_state[msg.cell_id] = msg.state;
+                }
+            }
+            // Compute next state
+            S next = local_computation();
+            // If next state is not the current state, then I change my state and schedule my next internal transition
+            if (next != state.current_state) {
+                buffer->add_to_buffer(next, clock + output_delay(next));
+                next_internal = buffer->next_timeout() - clock;
+            }
+            state.current_state = next;
+        }
+
+        /// Confluence transition function.
+        void confluence_transition(T e, typename cadmium::make_message_bags<input_ports>::type mbs) {
+            internal_transition();
+            external_transition(T(), move(mbs));
+        }
```

Comment:
Isn't this one to be expected to be fill by users?

Reply:
Not in [Cell-DEVS](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9004763&casa_token=4JRwIBxX-lMAAAAA:3wLEC3JuY80n-oweqfB_N1ZeWK3afLpJk0oXPUw4YhR1zEXDhZDj8vh_d_--t_7W9MYf4vpsaQ&tag=1). This formalism is on top of DEVS, and the user only has to define the local computation function (i.e., the next state of the cell depending on its previous state and its neighbors' state) and the delay function (i.e., how long the cell has to wait before communicating to its neighbors this new state). The classic DEVS functions are hidden from the user.

Reply:
I was implementing DEVS some time ago (you can check some of its code in the basic_models/devs/) but I never finished the engine part for it. Are you going to use DEVS or PDEVS?

Reply:
I think PDEVS. Now the behavior is simple, but I am planning to add additional ports, etc.